[Tan Yiming] iP

tanyyyming opened the pull request:
## J.A.R.V.I.S.
> Started out, J.A.R.V.I.S. was just a natural language UI. Now he runs the Iron Legion. He runs more of the business than anyone besides Pepper. " ―[Tony Stark](https://ironman.fandom.com/wiki/Tony_Stark_(film)) to Bruce Banner

**Just A Rather Very Intelligent System (J.A.R.V.I.S.)** was originally Tony Stark's natural-language user interface computer system, named after [Edwin Jarvis](https://ironman.fandom.com/wiki/Edwin_Jarvis_(film)). This version of J.A.R.V.I.S. is a *task-management chatbot*. It has:
- A basic GUI
- Customised response
- ~~fast~~ SUPER SLOW to use

### Features
1. Remember the tasks you input and save them for future sessions;
2. Mark and unmark tasks
3. List all tasks on the calendar
4. Search tasks by keywords

Yes, I know it's `LAME`. I totally agree 🤡.

### Current progress
- [x] Level 1 - 10
- [x] A basic tagging system
- [ ] Rank tasks by date

### Running
Launch the GUI of J.A.R.V.I.S. from here:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

tanyyyming committed:
feat: implement the initial skeletal version of the chatbot Jarvis

tanyyyming committed:
feat: add echo method

tanyyyming committed:
feat: personalise Jarvis

tanyyyming committed:
feat: add task saving feature

tanyyyming committed:
feat: add task completion marking/unmarking feature

tanyyyming committed:
feat: add todo list functionality

tanyyyming committed:
chore(text-ui-test): add annotations for test scripts

tanyyyming committed:
test: use I/O redirection to automate text-ui-test

tanyyyming committed:
refactor: add a command resolver class to resolve user input

tanyyyming committed:
feat: add customised business errors to Jarvis

tanyyyming committed:
feat: add IndexOutOfRange business error

tanyyyming committed:
feat: add delete tasks feature

tanyyyming committed:
Add the Storage class to store tasks in and read tasks from a file

tanyyyming committed:
Modify the ListManager class to read/write to the task list file

tanyyyming committed:
Refactor the Basic and Format class to Ui class

tanyyyming committed:
Merge branch 'branch-level-7'

tanyyyming committed:
Add date/time validation and formatting for deadline and event

tanyyyming committed:
Merge branch 'branch-Level-8'

tanyyyming committed:
Refactor code to follow OOP more closely

tanyyyming committed:
Add the find task by keyword feature

tanyyyming committed:
Merge branch 'add-gradle-support'

tanyyyming committed:
Decapitalize the package name Jarvis to jarvis

tanyyyming committed:
Add gradle support to build and run Jarvis

tanyyyming committed:
Test the execute method in Parser and add method in TaskList

tanyyyming committed:
Change the saving directory of the task file

tanyyyming committed:
Add the biz error about data file creation

tanyyyming committed:
Write Javadoc for most public methods

tanyyyming committed:
Merge branch 'branch-Level-9'

tanyyyming committed:
Add the find method to the ITaskList interface

tanyyyming committed:
Merge branch 'branch-A-JavaDoc'

tanyyyming committed:
Write Javadoc for the find method

tanyyyming committed:
Modify all methods of TaskList to return String

tanyyyming committed:
Initialize JacaFX GUI with FXML

tanyyyming committed:
Use Mockito for unit testing

tanyyyming committed:
Implement fully functioning GUI for Jarvis

tanyyyming committed:
Merge branch 'branch-Level-10'

tanyyyming committed:
Format code to conform to coding standard

tanyyyming committed:
Merge branch 'branch-A-CodingStandard'

tanyyyming committed:
Add CheckStyle to detect coding standard violations

tanyyyming committed:
Merge branch 'branch-A-CheckStyle'

tanyyyming committed:
Dissociate the TextUi class from TaskList class

tanyyyming committed:
Change the main class in build.gradle

nicolengk commented on src/main/java/jarvis/Jarvis.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Scanner;
+
+public class Jarvis {
+    public final static String JARVIS_LOGO = "                                                                 \n" +
```
Nice that you remembered to make constant names uppercase with underscores

tanyyyming replied:
Thanks!

nicolengk reviewed (commented):
Looks good to me, your code doesn't seem to violate any serious coding standards. Nice job 👍

bnahor commented on src/main/java/command/Parser.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    /**
+     * Executes the command with the given arguments.
```
Well structured! Do consider incorporating a default case

tanyyyming replied:
Got it!

bnahor reviewed (commented):
Great Job! Minor suggestion made to switch statement format

LHeng1 commented on src/main/java/services/TextUi.java:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * The horizontal line for formatting messages.
+     */
+    private final static String HORIZONTAL_LINE = "__________________________________________________"
+            + "_______________________\n";
```
I like how you put in the effort to ensure line length is shorter than 120 characters 👍

tanyyyming replied:
Thanks!

LHeng1 commented on src/main/java/services/tasklist/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+
+public class TaskList implements ITaskList {
+    /** The list of tasks. */
+    protected List<Task> taskList;
```
Perhaps use plural form for variable name since it is a collection of objects (A list of tasks)? 

An example from the [SE Coding Standard Guide](https://se-education.org/guides/conventions/java/intermediate.html#naming) is:
```
Collection<Point> points;
```
So for your case, perhaps you can use ```List<Task> tasks```, if it makes sense.

tanyyyming replied:
Ok! I'll take note of it and change it soon!

LHeng1 commented on src/main/java/services/tasklist/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+     * @throws IndexOutOfRangeException if the task number is out of range.
+     */
+    @Override
+    public String delete(int taskNumber) throws SaveToFileException, IndexOutOfRangeException {
```
Perhaps a more intuitive method name here would be `deleteTask`?

tanyyyming replied:
Totally agree! Will change the name of my add method also!

LHeng1 reviewed (commented):
Just a few nits on naming here and there, but overall LGTM as code is concise and easy to read!! 👍👍

tanyyyming committed:
Adjust GUI padding to be more readable

tanyyyming committed:
Use java assertions to detect potential bugs

The code does not have any assertions to validate some assumptions
made in the class TaskList and Storage.

This makes the code vulnerable to potential bugs and runtime
errors if the assumptions do not hold after some code change.

Let's validate the instantiation of the Storage class indeed
creates a valid file for data storage and the load method
can load the encoded tasks correctly.

Let's also make sure the description passed to the add
method of TaskList is indeed non-empty.

Doing so allows the two class to be less error-proned,
and we can be alerted to any errors if the assertions
fail.

tanyyyming committed:
Refactor code to improve code quality

The code's most case statements lack a default branch. The naming
of some variables and methods are also unclear. The execute method
in the Parser class is also very long.

These offenses make the code unsafe and decrease its readability.

Let's add default branches to explicitly state that the default
branch is never reached. Let's also rename some variables and
methods to explain their purpose more clearly. Let's finally
extract smaller methods from the long execute method while
following the SLAP.

By doing so, the code becomes more readable and robust.

tanyyyming committed:
Use streams to make methods more concise

tanyyyming committed:
Merge pull request #1 from tanyyyming/branch-A-CodeQuality

Refactor code to improve code quality

tanyyyming committed:
Merge branch 'master' into branch-A-Assertions

tanyyyming committed:
Merge pull request #2 from tanyyyming/branch-A-Assertions

Branch-A-Assertions

tanyyyming committed:
Merge branch 'master' into branch-A-Streams

tanyyyming committed:
Merge pull request #3 from tanyyyming/branch-A-Streams

Branch-A-Streams

tanyyyming committed:
Add basic tagging to tasks and support add/delete tags

tanyyyming committed:
Merge branch 'tagging'

tanyyyming committed:
Improve null handling on the Tag class

tanyyyming committed:
Add explicit annotations for error handling

tanyyyming committed:
Merge branch 'branch-A-MoreErrorHandling'

tanyyyming committed:
Delete unused text-ui

tanyyyming committed:
Add more JavaDoc

tanyyyming committed:
Merge branch 'tagging'

tanyyyming committed:
Add more unit tests

tanyyyming committed:
Increase readability of error messages

tanyyyming committed:
Change the shadowJar version to 0.2

tanyyyming committed:
Add the product name Jarvis on GUI window

tanyyyming committed:
Add a user guide for Jarvis

tanyyyming committed:
Allow GUI to display long text fully

tanyyyming committed:
Modify build.gradle to allow Windows OS compatibility

tanyyyming committed:
Refine the user guide

tanyyyming committed:
Improve GUI for better visual effect

tanyyyming committed:
Refactor the load method of Storage class to reduce its length

tanyyyming committed:
Refine the GUI

tanyyyming committed:
Change README.md
